velox/exec/MarkDistinct.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/MarkDistinct.h"
+#include "velox/common/base/Range.h"
+#include "velox/vector/FlatVector.h"
+
+#include <algorithm>
+#include <utility>
+
+namespace facebook::velox::exec {
+
+MarkDistinct::MarkDistinct(
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::MarkDistinctNode>& planNode)
+    : Operator(
+          driverCtx,
+          planNode->outputType(),
+          operatorId,
+          planNode->id(),
+          "MarkDistinct") {
+  const auto& inputType = planNode->sources()[0]->outputType();
+
+  // Set all input columns as identity projection.
+  for (auto i = 0; i < inputType->size(); ++i) {
+    identityProjections_.emplace_back(i, i);
+  }
+
+  // We will use result[0] for distinct mask output.
+  resultProjections_.emplace_back(0, inputType->size());
+
+  // Initialize groupingset
+  auto numHashers = planNode->distinctKeys().size();
+  std::vector<std::unique_ptr<VectorHasher>> hashers;
+  hashers.reserve(numHashers);
+
+  for (const auto& distinctVariable : planNode.get()->distinctKeys()) {
+    auto channel = exprToChannel(distinctVariable.get(), inputType);
+    VELOX_CHECK_NE(channel, kConstantChannel);
+    hashers.push_back(VectorHasher::create(distinctVariable->type(), channel));
+  }
+
+  groupingSet_ = GroupingSet::createForMarkDistinct(
+      std::move(hashers), operatorCtx_.get());
+
+  results_.resize(1);
+}
+
+void MarkDistinct::addInput(RowVectorPtr input) {
+  groupingSet_->addInput(input, false /*mayPushdown*/);
+
+  input_ = std::move(input);
+}
+
+RowVectorPtr MarkDistinct::getOutput() {
+  if (isFinished() || !input_) {
+    return nullptr;
+  }
+
+  auto outputSize = input_->size();
+  // Re-use memory for the ID vector if possible.
+  VectorPtr& result = results_[0];
+  if (result && result.unique()) {
```

Comment:
Please can you explain what this logic is.

Reply:
Does the comment on line 75 help?

Reply:
I understood what your intent is, but I couldn't follow why the code you wrote is sufficient for it. Please can you explain about it.

Reply:
Please state specifically feels insufficient so I may address it.